### PR TITLE
Asymmetric tokens

### PR DESCRIPTION
**Pull request by Eh2406**
Builds on and is blocked by #10592. This adds initial support for Asymmetric Tokens #10519.

### CODE REVIEW

**Comment by Eh2406** [comment]
Big shout out to @brycx for all the work getting `pasetors` ready for this PR!

**Comment by joshtriplett** [comment]
@Eh2406 I think we should make them the same feature, since we shouldn't ship authenticated HTTP registries with non-asymmetric tokens.

File: src/cargo/util/auth.rs [hunk 1]
```diff
@@ -228,46 +270,122 @@ my-registry = {{ index = "{}" }}
...
+                            .as_bytes(),
+                    ),
+                    None,
+                )?
+            }
+        })
+    }
+}
+
+pub struct Mutation<'a> {
```
Comment (jonhoo) [inline src/cargo/util/auth.rs:360]: Drive-by comment: this would probably be better as an `enum`.

**Comment by Eh2406** [comment]
I just noticed that we have a check for setting the token using the command line, this check should probably be extended to the secret key as well.
https://github.com/rust-lang/cargo/blob/247ca7fd04e5c7cc383bcac7d72316e98fbea2bc/src/cargo/util/config/mod.rs#L1264

**Comment by ehuss** [comment]
With #10592 landed, is there anything needed to be ready for review other than resolving the merge conflicts?

**Comment by Eh2406** [comment]
It needs to be rebased and updated. I remember that there was significant feedback, but don't remember what it was offhand. I'm planning to come back to this PR the end of this week.

File: src/cargo/sources/registry/http_remote.rs [hunk 2]
```diff
@@ -591,9 +591,14 @@ impl<'cfg> RegistryData for HttpRegistry<'cfg> {
...
-                auth::auth_token(self.config, &self.source_id, self.login_url.as_ref())?;
-            headers.append(&format!("Authorization: {}", authorization))?;
+            let credential = auth::auth_token(self.config, &self.source_id)?;
+            let authorization = credential.as_header(
+                self.config,
+                &self.source_id,
+                self.login_url.as_ref(),
+                None,
+            )?;
+            headers.append(&format!("authorization: {}", authorization))?;
```
Comment (arlosi) [inline src/cargo/sources/registry/http_remote.rs:601]: Keep the capitalized `Authorization`, otherwise the log-redaction doesn't work.

```suggestion
            headers.append(&format!("Authorization: {}", authorization))?;
```

File: src/cargo/util/auth.rs [hunk 3]
```diff
@@ -133,52 +158,71 @@ pub fn registry_credential_config(
...
+    let err_both = |token_key: &str, proc_key: &str| {
+        Err(format_err!(
+            "both `{token_key}` and `{proc_key}` \
+            were specified in the config for registry `{name}`.\n\
+             Only one of these values may be set, remove one or the other to proceed.",
+            name = name.unwrap()
+        ))
+    };
+    Ok(
+        match (token, credential_process, secret_key, secret_key_subject) {
```
Comment (arlosi) [inline src/cargo/util/auth.rs:190]: Since this has gotten more complicated, it may be worth deduplicating with the crates.io case above.

File: crates/cargo-test-support/src/registry.rs [hunk 4]
```diff
@@ -650,9 +698,139 @@ impl HttpServer {
...
+            pub name: Option<&'a str>,
+            pub vers: Option<&'a str>,
+            pub cksum: Option<&'a str>,
+        }
+        let authorized = |mutation: Option<Mutation>| {
+            if mutation.is_some() || self.auth_required {
+                match &self.token {
+                    Token::Bearer(token) => Some(token) == req.authorization.as_ref(),
+                    Token::Keys(private_key, private_key_subject) => {
+                        macro_rules! t {
```
Comment (arlosi) [inline crates/cargo-test-support/src/registry.rs:712]: Consider factoring this out into a separate method for validating the PASETO

File: crates/cargo-test-support/src/registry.rs [hunk 5]
```diff
@@ -55,12 +60,30 @@ fn generate_url(name: &str) -> Url {
     Url::from_file_path(generate_path(name)).ok().unwrap()
 }
 
+#[derive(Clone)]
+pub enum Token {
+    Bearer(String),
```
Comment (arlosi) [inline crates/cargo-test-support/src/registry.rs:65]: It's not really a bearer token, since Cargo/crates.io made an implementation error and don't actually prefix the token with `Bearer `.

File: crates/cargo-test-support/src/registry.rs [hunk 6]
```diff
@@ -118,7 +149,7 @@ impl RegistryBuilder {
     pub fn new() -> RegistryBuilder {
         RegistryBuilder {
             alternative: None,
-            token: None,
+            token: Token::Bearer("api-token".to_string()),
```
Comment (arlosi) [inline crates/cargo-test-support/src/registry.rs:152]: Why this change? Previously a value of `None` would be changed to the default of `"{prefix}sekrit"`.

**Review by arlosi — COMMENTED** [review]
Could you also update `unstable.md` to describe the feature?

File: src/cargo/sources/registry/http_remote.rs (hunk 2 above)
Comment (Eh2406) [reply src/cargo/sources/registry/http_remote.rs:601]: Oops.. will be fixed in the next push.

File: crates/cargo-test-support/src/registry.rs (hunk 6 above)
Comment (Eh2406) [reply crates/cargo-test-support/src/registry.rs:152]: Good catch. Will be fixed in the next push.

File: crates/cargo-test-support/src/registry.rs (hunk 5 above)
Comment (Eh2406) [reply crates/cargo-test-support/src/registry.rs:65]: Will be fixed in the next push.

File: crates/cargo-test-support/src/registry.rs (hunk 4 above)
Comment (Eh2406) [reply crates/cargo-test-support/src/registry.rs:715]: done

**Comment by Eh2406** [comment]
This is now ready for review. I have tried to follow-up on all feedback that was provided.

There are some things that were left for follow-up PR's:
- Making a `keep this secure` wrapper around various strings, so that the type system marks that they should not be printed. Specifically this should wrap plaintext tokens and PASETO after they've been generated. An audit was done to make sure no structure containing the data implements `Display`, nor is currently printed. But, moving it to the type system was left for later.
- Support for asymmetric tokens being returned from credential processes.
- Support for challenges from the server.

File: tests/testsuite/registry.rs [hunk 7]
```diff
@@ -1127,6 +1127,95 @@ fn login_with_token_on_stdin() {
...
+    let credentials = paths::home().join(".cargo/credentials");
+    fs::remove_file(&credentials).unwrap();
+    cargo_process("login --key-subject=foo -Z registry-auth")
+        .masquerade_as_nightly_cargo(&["registry-auth"])
+        .replace_crates_io(registry.index_url())
+        .with_stderr_contains("error: need a secret_key to set a key_subject")
+        .with_status(101)
+        .run();
+
+    // ok so ad a secret_key to the credentials
```
Comment (ehuss) [inline tests/testsuite/registry.rs:1182]:
```suggestion
    // ok so add a secret_key to the credentials
```

**Review by ehuss — COMMENTED** [review]
This is just a partial review, will try to finish the rest later.

File: tests/testsuite/registry.rs [hunk 8]
```diff
@@ -1127,6 +1127,95 @@ fn login_with_token_on_stdin() {
...
+        .replace_crates_io(registry.index_url())
+        .with_stdout(
+            "\
+        please paste the API secret key below
+k3.public.AmDwjlyf8jAV3gm5Z7Kz9xAOcsKslt_Vwp5v-emjFzBHLCtcANzTaVEghTNEMj9PkQ",
+        )
+        .with_stdin("k3.secret.fNYVuMvBgOlljt9TDohnaYLblghqaHoQquVZwgR6X12cBFHZLFsaU3q7X3k1Zn36")
+        .run();
+    let credentials = fs::read_to_string(&credentials).unwrap();
+    assert!(credentials.starts_with("[registry]\n"));
```
[the remark this answers is not in the document]
Comment (Eh2406) [reply tests/testsuite/registry.rs:1146]: What order the keys are serialized is not deterministic. If the order is hardcoded, then the test will fail half the time.

File: src/cargo/ops/registry.rs [hunk 9]
```diff
@@ -750,48 +804,95 @@ pub fn registry_login(config: &Config, token: Option<&str>, reg: Option<&str>) -
...
-            line.replace("cargo login", "").trim().to_string()
+            secret_key = line.trim().to_string();
+        } else {
+            secret_key = old_secret_key
+                .cloned()
+                .ok_or_else(|| anyhow!("need a secret_key to set a key_subject"))?;
         }
-    };
+        if let Some(p) = paserk_public_from_paserk_secret(&secret_key) {
+            drop_println!(config, "{}", &p);
```
[the remark this answers is not in the document]
Comment (Eh2406) [reply src/cargo/ops/registry.rs:843]: I had kind of forgotten about this functionality until re-reading the RFC yesterday. The RFC describes how after generating a secret key, `cargo login` will print the public key to give to the registry. We can print with more context/instructions. By putting it directly we make it easier for this to be hooked into automation. Suggestions are definitely welcome.

**Comment by Eh2406** [comment]
Thanks for the review. I took care of a bunch of the smaller items. I will work on larger items tomorrow.

File: src/cargo/ops/registry.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (ehuss) [reply src/cargo/ops/registry.rs:847]: One option is to do something like what `cargo vendor` does, where it prints the instructions on stderr, and anything that is needed for output on stdout.

However, in this case, I'm not clear what kind of automation there would be. Particularly if the secret key has to be pasted via a dialog.  If this public key is needed for something, perhaps there could be some kind of separate command to print it? I'm not really clear what the use case is here, though.

File: src/cargo/ops/registry.rs [hunk 10]
```diff
@@ -750,48 +804,95 @@ pub fn registry_login(config: &Config, token: Option<&str>, reg: Option<&str>) -
...
-            } else {
-                drop_println!(
-                    config,
-                    "please paste the token for {} below",
-                    source_ids.original.display_registry_name()
-                )
+    let new_token;
+    if generate_keypair || secret_key_required || key_subject.is_some() {
+        if !config.cli_unstable().registry_auth {
+            panic!("-Zregistry_auth required.");
```
[the remark this answers is not in the document]
Comment (Eh2406) [reply src/cargo/ops/registry.rs]: Looking more closely `fail_if_stable_opt` is hardcoded to check `self.unstable_options` https://github.com/rust-lang/cargo/blob/cc0a320879c17207bbfb96b5d778e28a2c62030d/src/cargo/core/features.rs#L989 if I add an argument like `zflage: &str` I don't see a good way to check if that flag has been set.

[the remark this answers is not in the document]
Comment (ehuss) [reply src/cargo/ops/registry.rs]: Yea, it doesn't fit with the current design. I'm happy just adding a little extra text for now. This is a relatively unusual situation, which we can examine more if it comes up again in the future.

[the remark this answers is not in the document]
Comment (Eh2406) [reply src/cargo/ops/registry.rs]: Just push changes that improve the text to better match `fail_if_stable_opt` and added a test.

File: tests/testsuite/login.rs [hunk 11]
```diff
@@ -123,3 +123,33 @@ fn empty_login_token() {
         .with_status(101)
         .run();
 }
+
+#[cargo_test]
+fn bad_asymmetric_token_args() {
```
[the remark this answers is not in the document]
Comment (Eh2406) [reply tests/testsuite/login.rs:128]: Test added.

File: src/cargo/ops/registry.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (Eh2406) [reply src/cargo/ops/registry.rs:856]: I don't know what the long-term best "correct interface" looks like. This is probably not the best place to have a full discussion of all the possible use cases and their ergonomics.
I suggest we merge *something* here, and add a follow-up item to design the correct interface.
I would be comfortable with merging any of:
- do not print it
- print the public key with human readable instructions
- split stout and sterr like `cargo vendor`

But I would like to get this PR merged before we get too far into designing a "correct interface". What would you prefer?

File: tests/testsuite/registry.rs [hunk 12]
```diff
@@ -1127,6 +1127,95 @@ fn login_with_token_on_stdin() {
     assert_eq!(credentials, "[registry]\ntoken = \"some token\"\n");
 }
 
+#[cargo_test]
+fn login_with_asymmetric_token_and_subject_on_stdin() {
```
[the remark this answers is not in the document]
Comment (Eh2406) [reply tests/testsuite/registry.rs]: moved

File: src/cargo/util/config/mod.rs [hunk 13]
```diff
@@ -193,7 +193,7 @@ pub struct Config {
     updated_sources: LazyCell<RefCell<HashSet<SourceId>>>,
     /// Cache of credentials from configuration or credential providers.
     /// Maps from url to credential value.
-    credential_cache: LazyCell<RefCell<HashMap<CanonicalUrl, String>>>,
+    credential_cache: LazyCell<RefCell<HashMap<CanonicalUrl, (bool, String)>>>,
```
[the remark this answers is not in the document]
Comment (Eh2406) [reply src/cargo/util/config/mod.rs:196]: done

File: tests/testsuite/registry_auth.rs [hunk 14]
```diff
@@ -64,6 +64,19 @@ fn simple() {
...
+fn simple_with_asymmetric() {
+    let _registry = RegistryBuilder::new()
+        .alternative()
+        .auth_required()
+        .http_index()
+        .token(cargo_test_support::registry::Token::rfc_key())
+        .build();
+
+    let p = make_project();
+    cargo(&p, "build").with_stderr(SUCCCESS_OUTPUT).run();
```
[the remark this answers is not in the document]
Comment (Eh2406) [reply tests/testsuite/registry_auth.rs:77]: Added tests.

**Comment by Eh2406** [comment]
I have addressed or responded to your first partial review.

File: src/cargo/util/auth.rs [hunk 15]
```diff
@@ -273,37 +330,168 @@ pub fn auth_token(config: &Config, sid: &SourceId, login_url: Option<&Url>) -> C
...
+    #[serde(skip_serializing_if = "Option::is_none")]
+    name: Option<&'a str>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    vers: Option<&'a str>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    cksum: Option<&'a str>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    challenge: Option<&'a str>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    v: Option<u8>,
```
Comment (ehuss) [inline src/cargo/util/auth.rs:473]: The `v` field does not seem to be used. I'm inferring this would only be set in the future (to something like `Some(2)`) if there is a breaking change? Can some documentation about this be added here?

File: src/cargo/util/auth.rs [hunk 16]
```diff
@@ -273,37 +330,168 @@ pub fn auth_token(config: &Config, sid: &SourceId, login_url: Option<&Url>) -> C
...
 }
 
 /// Saves the given token.
-pub fn login(config: &Config, sid: &SourceId, token: String) -> CargoResult<()> {
+pub fn login(config: &Config, sid: &SourceId, token: RegistryCredentialConfig) -> CargoResult<()> {
     match registry_credential_config(config, sid)? {
         RegistryCredentialConfig::Process(process) => {
+            let token = token
+                .as_token()
+                .expect("credential_process can not use login with a secret_key")
```
Comment (ehuss) [inline src/cargo/util/auth.rs:493]:
```suggestion
                .expect("credential_process cannot use login with a secret_key")
```

File: crates/cargo-test-support/src/registry.rs [hunk 17]
```diff
@@ -287,32 +320,48 @@ impl RegistryBuilder {
...
+            match &registry.token {
+                Token::Plaintext(token) => {
+                    if let Some(alternative) = &self.alternative {
+                        append(
+                            &credentials,
+                            format!(
+                                r#"
                     [registries.{alternative}]
                     token = "{token}"
                 "#
```
[the remark this answers is not in the document]
Comment (Eh2406) [reply crates/cargo-test-support/src/registry.rs:333]: Sorry, I thought I had fixed this before I marked resolved. I will merge your suggestion.

File: src/cargo/ops/registry.rs [hunk 18]
```diff
@@ -148,14 +167,18 @@ pub fn publish(ws: &Workspace<'_>, opts: &PublishOpts<'_>) -> CargoResult<()> {
             );
         }
     }
+    // This is only used to confirm that we can create a token before we build the package.
+    // This causes the credential provider to be called an extra time, but keeps the same order of errors.
+    let ver = pkg.version().to_string();
+    let mutation = auth::Mutation::PrePublish;
```
[the remark this answers is not in the document]
Comment (Eh2406) [reply src/cargo/ops/registry.rs:173]: I added the commit that caches more aggressively. When we implement asymmetric tokens with credential-process, the `run_command` will need to report on if it looked at the mutation. When the mutation is not consulted by a credential-process, then the token can be cashed.

File: src/cargo/util/auth.rs [hunk 19]
```diff
@@ -133,52 +134,99 @@ pub fn registry_credential_config(
...
+    were specified in the config{registry}.\n\
+        Only one of these values may be set, remove one or the other to proceed.",
+        ))
+    };
+    Ok(
+        match (token, credential_process, secret_key, secret_key_subject) {
+            (Some(_), Some(_), _, _) => return err_both("token", "credential-process"),
+            (Some(_), _, Some(_), _) => return err_both("token", "secret-key"),
+            (_, Some(_), Some(_), _) => return err_both("credential-process", "secret-key"),
+            (_, _, None, Some(_)) => {
```
[the remark this answers is not in the document]
Comment (Eh2406) [reply src/cargo/util/auth.rs:193]: One is tested [here](https://github.com/rust-lang/cargo/blob/9cb39418a57ec7cd833b95eac16a9d89c570df2a/tests/testsuite/credential_process.rs#L70), I will add tests for the other 2.

**Comment by Eh2406** [comment]
I believe this is ready for another round of review.

File: src/cargo/ops/registry.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (ehuss) [reply src/cargo/ops/registry.rs:173]: I'm not quite following how the commit is supposed to help avoid fetching the token twice.

From what I can tell, `auth_token_optional` will only insert into the cache if `mutation.is_none()`. However, when publishing, the first pass has a `mutation` of `Some(PrePublish)` and the second pass has a mutation of `Some(Publish)`, so neither case will cache it. The new `independent_of_endpoint` value doesn't seem to come into play. But even if it did, it seems to be forced to be `false` which means it won't force the cache to be used.

**Comment by ehuss** [comment]
🎉  Thanks! I appreciate that you clearly put in a lot of effort on this.

Can you collect any todo items, and create new issues for them? I think it would help to track what is left, what needs to be decided or resolved, and to organize discussion on those.

@bors r+

**Comment by Eh2406** [comment]
That seams spurious and unrelated.
